### PR TITLE
Adding Document Sets

### PR DESCRIPTION
**Pull request by Weves**
Adds:
- name for connector credential pairs + frontend changes to start populating this field
- document set table migration
- during indexing, document sets are now checked and inserted into Vespa
- background job to check if document sets need to be synced
- document set management APIs
- document set management dashboard in the UI

### CODE REVIEW

File: backend/alembic/versions/800f48024ae9_add_id_to_connectorcredentialpair.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        ),
+    )
+    op.add_column(
+        "connector_credential_pair",
+        sa.Column("name", sa.String(), nullable=True),
+    )
+
+    # fill in IDs for existing rows
+    op.execute(
+        "UPDATE connector_credential_pair SET id = nextval('connector_credential_pair_id_seq') WHERE id IS NULL"
```
Comment (yuhongsun96) [inline backend/alembic/versions/800f48024ae9_add_id_to_connectorcredentialpair.py:35]: How could this not be null?

File: backend/danswer/background/celery/deletion_utils.py [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+    error_msg = None
+    num_docs_deleted = 0
+    if deletion_task_status == "SUCCESS":
+        deletion_status = DeletionStatus.SUCCESS
+        num_docs_deleted = cast(int, deletion_task.get(propagate=False))
+    elif deletion_task_status == "FAILURE":
+        deletion_status = DeletionStatus.FAILED
+        error_msg = deletion_task.get(propagate=False)
+    elif deletion_task_status == "STARTED" or deletion_task_status == "PENDING":
+        deletion_status = DeletionStatus.IN_PROGRESS
```
Comment (yuhongsun96) [inline backend/danswer/background/celery/deletion_utils.py:29]: Should we sync all our statuses to Celery status?

Also why do we have different ones for Deletion and others :P Could just have a job status enum I guess no?

File: backend/danswer/background/document_set_sync_script.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
+from celery.result import AsyncResult
```
Comment (yuhongsun96) [inline backend/danswer/background/document_set_sync_script.py:1]: preference to put this under the backend/scripts directory and use from there

File: backend/danswer/background/document_set_sync_script.py [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+                logger.info(
+                    f"Document set {document_set.id} is not synced. Syncing now!"
+                )
+                task = sync_document_set_task.apply_async(
+                    kwargs=dict(document_set_id=document_set.id),
+                )
+                _ExistingTaskCache[document_set.id] = task
+
+
+if __name__ == "__main__":
```
Comment (yuhongsun96) [inline backend/danswer/background/document_set_sync_script.py:50]: Can we use something like Celery Beat? Would be great to port the Indexing jobs over as well. The idea is to not having all these different files registering their own celery jobs every so often and doing nothing else.

I can take it up later, don't consider this blocking

File: backend/danswer/background/utils.py [hunk 5]
```diff
@@ -8,10 +8,13 @@
 logger = setup_logger()
 
 
-def interval_run_job(job: Callable[[], Any], delay: int | float) -> None:
+def interval_run_job(
```
Comment (yuhongsun96) [inline backend/danswer/background/utils.py:11]: Celery Beat angle for sure

Comment (yuhongsun96) [reply backend/danswer/background/utils.py:11]: app.conf.beat_schedule = {
    'my-task': {
        'task': 'myapp.tasks.my_periodic_task',
        'schedule': timedelta(minutes=5),  # Execute every 5 minutes
    },
}

something like this, centralized and CLEAN!

File: backend/danswer/db/document_set.py [hunk 6]
```diff
@@ -0,0 +1,282 @@
...
+from danswer.db.models import ConnectorCredentialPair
+from danswer.db.models import Document
+from danswer.db.models import DocumentByConnectorCredentialPair
+from danswer.db.models import DocumentSet as DocumentSetDBModel
+from danswer.db.models import DocumentSet__ConnectorCredentialPair
+from danswer.server.models import DocumentSetCreationRequest
+from danswer.server.models import DocumentSetUpdateRequest
+
+
+def _delete_document_set_cc_pairs(
```
Comment (yuhongsun96) [inline backend/danswer/db/document_set.py:20]: Can we come up with a name for this? I called it something something no commit at one point, lets have it in the function so there's not a bug that crops up later

File: backend/danswer/db/document_set.py [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+        db_session.add_all(ds_cc_pairs)
+        db_session.commit()
+    except:
+        db_session.rollback()
+        raise
+
+    return new_document_set_row, ds_cc_pairs
+
+
+def update_document_set(
```
Comment (yuhongsun96) [inline backend/danswer/db/document_set.py:88]: Makes sense but a comment on the update document set + deleting expired doc set cc pair logic would be nice

File: backend/danswer/db/models.py [hunk 8]
```diff
@@ -326,6 +339,49 @@ class Document(Base):
...
+class DocumentSet__ConnectorCredentialPair(Base):
+    __tablename__ = "document_set__connector_credential_pair"
+
+    document_set_id: Mapped[int] = mapped_column(
+        ForeignKey("document_set.id"), primary_key=True
+    )
+    connector_credential_pair_id: Mapped[int] = mapped_column(
+        ForeignKey("connector_credential_pair.id"), primary_key=True
+    )
+    # if `True`, then is part of the current state of the document set
```
Comment (yuhongsun96) [inline backend/danswer/db/models.py:368]: What was the final reasoning of using is_current vs versions? Simplicity?

File: backend/danswer/db/document_set.py [hunk 9]
```diff
@@ -0,0 +1,282 @@
...
+    if current_only:
+        stmt = stmt.where(
+            DocumentSet__ConnectorCredentialPair.is_current == True  # noqa: E712
+        )
+    stmt = stmt.distinct()
+
+    return db_session.scalars(stmt).all()
+
+
+def fetch_document_sets_for_documents(
```
Comment (yuhongsun96) [inline backend/danswer/db/document_set.py:250]: Let's discuss how we can specify user access to document sets, right now everything is fully open and filtered at the Vespa level by ACL right?

**Review by yuhongsun96 — APPROVED** [review]

File: backend/alembic/versions/800f48024ae9_add_id_to_connectorcredentialpair.py (hunk 1 above)
Comment (Weves) [reply backend/alembic/versions/800f48024ae9_add_id_to_connectorcredentialpair.py:35]: If the migration is run twice for some reason

File: backend/danswer/background/document_set_sync_script.py (hunk 4 above)
Comment (Weves) [reply backend/danswer/background/document_set_sync_script.py:50]: Sounds good!

File: backend/danswer/db/document_set.py (hunk 6 above)
Comment (Weves) [reply backend/danswer/db/document_set.py:20]: Let's do `__no_comit`

**Pull request merged**